lib/spack/spack/stage.py

```diff
@@ -855,6 +861,62 @@ def cache_local(self):
         tty.debug("Sources for DIY stages are not cached")
 
 
+class DevelopStage(Stage):
```

Comment:
This does seems like a lot of extra machinery. Also, how will this impact `spack cd`?  We use that extensively for navigating around in development environments and having the default behavior. Seems like this PR will change that behavior and send you to the log directory rather than the source directory for dev specs.

Reply:
> Also, how will this impact spack cd? 

Before this PR, `spack cd` would take you to the `dev_path` (your source)

With this PR, `spack cd` would take you to a parent of the source (`spack-stage-<hash>` from https://github.com/spack/spack/pull/41373#issuecomment-1945174065)

In a sense, that makes the behavior of `spack cd` consistent for develop and non-develop specs. `spack cd --source-dir` would take you to the `dev_path` directly though.